### PR TITLE
Document metrics and consumer tuning based on metrics

### PR DESCRIPTION
**Pull request by erikvanoosten**
Also: fix typo and make metric descriptions consistent.

### CODE REVIEW

File: docs/consumer-internals.svg
Comment (svroonland) [inline docs/consumer-internals.svg:1]: Can we include the prefetch buffer in this picture?

File: docs/consumer-tuning.md [hunk 1]
```diff
@@ -90,3 +90,37 @@ On older zio-kafka versions `withMaxPollInterval` is not available. Use the foll
...
 When no stream is processing data for this amount of time (while new data is available), the consumer will halt with a
 failure. In zio-kafka 2.5.0 `runloopTimeout` defaults to 4 minutes, a little bit lower than `max.poll.interval.ms`.
+
+## Using metrics to tune the consumer
+
+Zio-Kafka exposes [metrics](metrics.md) that can be used to further tune the consumer. To interpret these metrics you need to know how zio-kafka works internally.
+
+![](consumer-internals.svg)
+
+The runloop is at the heart of every zio-kakfa consumer.
```
Comment (svroonland) [inline docs/consumer-tuning.md:100]: typo

File: docs/metrics.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+## Zio-kafka consumer metrics
+
+The zio-kafka consumer collects some additional metrics using the zio-metrics API. This allows any zio-metrics backend
+to access and process the observed values.
+
+By default, no tags are added, but this can be configured via the new method `ConsumerSettings.withMetricsLabels`.
+
+The following metrics are collected:
+
+- Poll metrics: poll count (counter), number of records per poll (histogram), poll latency (histogram).
```
Comment (svroonland) [inline docs/metrics.md:24]: What about a table listing the exact metric names and a description?

File: docs/consumer-tuning.md [hunk 3]
```diff
@@ -90,3 +90,37 @@ On older zio-kafka versions `withMaxPollInterval` is not available. Use the foll
...
+It creates a zstream for each partition, eventually this is the zstream your applications consumes from.
+When the zstream starts, and every time the record queue is empty, it sends a request for data to the runloop.
+The request causes the runloop to resume the partition so that the next poll may receive records.
+Any received records are put in a record queue.
+When the record queue reaches a certain size (as determined by the configured `FetchStrategy`), the partition is paused.
+The zstream reads from the queue and emits the records to the user code.
+
+An optimally configured consumer has the following properties:
```
Comment (svroonland) [inline docs/consumer-tuning.md:108]: We should specify what it is optimal towards

File: docs/consumer-tuning.md (hunk 1 above)
Comment (erikvanoosten) [reply docs/consumer-tuning.md]: Fixed.

File: docs/consumer-tuning.md (hunk 3 above)
Comment (erikvanoosten) [reply docs/consumer-tuning.md:108]: Fixed.

File: docs/metrics.md (hunk 2 above)
Comment (erikvanoosten) [reply docs/metrics.md:24]: Sounds like a lot of work. And it might go out of date. WDYT of pointing to the code?

Comment (erikvanoosten) [reply docs/metrics.md]: Did it anyway.

File: docs/metrics.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+| counter   | `ziokafka_consumer_poll_auth_errors`                  | The number of polls that ended with an authentication or authorization error.      |
+
+### Partition stream metrics
+
+These metrics are updated after every poll.
+
+| Type      | Name                                 | Description                                                            |
+|-----------|--------------------------------------|------------------------------------------------------------------------|
+| histogram | `ziokafka_consumer_pending_requests` | The number of partitions that ran out of records (the queue is empty). |
+| histogram | `ziokafka_consumer_queue_size`       | The number of records queued for a partition.                          |
```
Comment (svroonland) [inline docs/metrics.md:49]: Should mention that these are tagged with topic and partition

**Review by svroonland — APPROVED** [review]
A nitpick, but otherwise a nice addition to the docs!

File: docs/metrics.md (hunk 4 above)
Comment (erikvanoosten) [reply docs/metrics.md:49]: That is a misunderstanding. They are not tagged, except with whatever is configured through the consumer settings.

**Pull request merged**